catalogue/webapp/components/Auth/Auth.js

```diff
@@ -0,0 +1,167 @@
+// @flow
+import Router from 'next/router';
+import { type Node, useEffect, useState } from 'react';
+import fetch from 'isomorphic-unfetch';
+import uuidv4 from 'uuid/v4';
+import crypto from 'crypto-browserify';
+import base64url from 'base64url';
+
+type AuthState = 'loggedOut' | 'authorising' | 'loggedIn' | 'expired';
+
+const authDomain = 'https://id.wellcomecollection.org';
+const authParams = {
+  response_type: 'code',
+  client_id: '5n4vt54rjsg6t691c5b5kiacdv',
+  scope: ['openid'].join(' '),
+};
+
+function getFromLocalStorage(key: string) {
+  return window.localStorage.getItem(key);
+}
+
+const authStates = {
+  loggedOut: 'loggedOut',
+  authorising: 'authorising',
+  loggedIn: 'loggedIn',
+  expired: 'expired',
+};
+
+function getLocalStorageJson(key: string) {
+  try {
+    return JSON.parse(getFromLocalStorage(key));
+  } catch (e) {
+    return null;
+  }
+}
+
+function getRedirectUrl() {
+  return `${window.location.href.split('/')[0]}//${
+    window.location.href.split('/')[2]
+  }/works/auth-code`;
+}
+
+async function getToken(code: string, verifier: string) {
+  const url = `${authDomain}/oauth2/token`;
+  const params = {
+    grant_type: 'authorization_code',
+    client_id: authParams.client_id,
+    redirect_uri: getRedirectUrl(),
+    code_verifier: verifier,
+    code: code,
+  };
+  const token = await fetch(url, {
```

Comment:
nit: I believe you could just `return fetch(...).then(...)` here, no need for the async/await

Reply:
I _think_ we want to return the resolved value, rather than the Promise? So at best it would have to be `return await…` (but I think we found out through eslint shouting at us that `return await`ing is discouraged).

Reply:
Async functions always return promises though - I believe `return await` is exactly equivalent to returning a promise, which is [why eslint doesn't like it](https://github.com/eslint/eslint/blob/master/docs/rules/no-return-await.md)

Reply:
To be clear though I don't think it matters all that much either way! Just thought I'd point it out

Reply:
Mainly done to use a `const` telling you what the promise is, although that is captured in the function name.